src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -90,4 +90,189 @@ bool IPhysicalDevice::validateLogicalDeviceCreation(const ILogicalDevice::SCreat
+
+    return formatBufUsg;
+}
+
+asset::E_FORMAT IPhysicalDevice::promoteBufferFormat(const FormatPromotionRequest<asset::IBuffer::E_USAGE_FLAGS> req)
+{
+    auto buf_cache = this->m_formatPromotionCache.buffers;
+    auto cached = buf_cache.find(req);
+    if (cached != buf_cache.end())
+        return cached->second;
+
+    auto srcUsages = convBufferUsage(req.usages);
+
+    // Cache valid formats per usage?
+    core::bitflag<asset::E_FORMAT> validFormats;
+
+    for (uint32_t format = 0; format < asset::E_FORMAT::EF_UNKNOWN; format++)
+    {
+        auto f = static_cast<asset::E_FORMAT>(format);
+        auto formatUsages = this->getBufferFormatUsages(f);
+        if ((srcUsages & formatUsages) == srcUsages)
+        {
+            validFormats |= f;
+        }
+    }
+
+    //if (validFormats.hasFlags(req.originalFormat)) return req.originalFormat;
+
+    return narrowDownFormatPromotion(validFormats, req.originalFormat);
+}
+
+video::IPhysicalDevice::SFormatImageUsage convImageUsage(core::bitflag<asset::IImage::E_USAGE_FLAGS> usages)
+{
+    video::IPhysicalDevice::SFormatImageUsage formatImgUsg;
+    formatImgUsg.isInitialized = 1;
+    formatImgUsg.sampledImage = usages.hasFlags(asset::IImage::EUF_SAMPLED_BIT);
+    formatImgUsg.storageImage = usages.hasFlags(asset::IImage::EUF_STORAGE_BIT);
+    formatImgUsg.transferSrc = usages.hasFlags(asset::IImage::EUF_TRANSFER_SRC_BIT);
+    formatImgUsg.transferDst = usages.hasFlags(asset::IImage::EUF_TRANSFER_DST_BIT);
+    formatImgUsg.attachment = (usages & core::bitflag<asset::IImage::E_USAGE_FLAGS>(
+        asset::IImage::EUF_COLOR_ATTACHMENT_BIT | asset::IImage::EUF_DEPTH_STENCIL_ATTACHMENT_BIT)).value != 0;
+    // Special flags for these?
+    formatImgUsg.blitSrc = usages.hasFlags(asset::IImage::EUF_TRANSFER_SRC_BIT);
+    formatImgUsg.blitDst = usages.hasFlags(asset::IImage::EUF_TRANSFER_DST_BIT);
+    formatImgUsg.storageImageAtomic = usages.hasFlags(asset::IImage::EUF_STORAGE_BIT);
+    formatImgUsg.attachmentBlend = (usages & core::bitflag<asset::IImage::E_USAGE_FLAGS>(
+        asset::IImage::EUF_COLOR_ATTACHMENT_BIT | asset::IImage::EUF_DEPTH_STENCIL_ATTACHMENT_BIT)).value != 0;
+
+    return formatImgUsg;
+}
+
+asset::E_FORMAT IPhysicalDevice::promoteImageFormat(const FormatPromotionRequest<asset::IImage::E_USAGE_FLAGS> req, const asset::IImage::E_TILING tiling)
+{
+    format_image_cache_t cache;
+    switch (tiling)
+    {
+    case asset::IImage::E_TILING::ET_LINEAR:
+        cache = this->m_formatPromotionCache.linearTilingImages;
+        break;
+    case asset::IImage::E_TILING::ET_OPTIMAL:
+        cache = this->m_formatPromotionCache.optimalTilingImages;
+        break;
+    default:
+        return asset::E_FORMAT::EF_UNKNOWN;
+    }
+    auto cached = cache.find(req);
+    if (cached != cache.end())
+        return cached->second;
+    auto srcUsages = convImageUsage(req.usages);
```

Comment:
do an early out https://github.com/Devsh-Graphics-Programming/Nabla/pull/375/files#r912026997, if you can use the format right away
```cpp
if (req.canUseFormat(getBufferFormatUsages(srcFormat))
{
   cache.insert(cached,{req,srcFormat});
   return srcFormat;
}
```